subsys/usb/class/audio/usb_audio_internal.h

```diff
@@ -0,0 +1,506 @@
+	.bDescriptorType = USB_CS_INTERFACE_DESC,			      \
+	.bDescriptorSubtype = USB_AUDIO_HEADER,				      \
+	.bcdADC = sys_cpu_to_le16(0x0100),				      \
+	.wTotalLength = sys_cpu_to_le16(				      \
+		sizeof(struct dev##_cs_ac_interface_descriptor_header_##i) +  \
+		INPUT_TERMINAL_DESC_SIZE +				      \
+		sizeof(struct dev##_feature_unit_descriptor_##i) +	      \
+		OUTPUT_TERMINAL_DESC_SIZE),				      \
+	.bInCollection = ifaces,					      \
+	.baInterfaceNr = {0},						      \
+}
+
+#define INIT_CS_AC_IF_HEADER_BIDIR(dev, i, ifaces)			      \
+{									      \
+	.bLength = sizeof(struct dev##_cs_ac_interface_descriptor_header_##i),\
+	.bDescriptorType = USB_CS_INTERFACE_DESC,			      \
+	.bDescriptorSubtype = USB_AUDIO_HEADER,				      \
+	.bcdADC = sys_cpu_to_le16(0x0100),				      \
+	.wTotalLength = sys_cpu_to_le16(				      \
+		sizeof(struct dev##_cs_ac_interface_descriptor_header_##i) +  \
+		2*INPUT_TERMINAL_DESC_SIZE +				      \
+		sizeof(struct dev##_MIC_feature_unit_descriptor_##i) +	      \
+		sizeof(struct dev##_HP_feature_unit_descriptor_##i) +	      \
+		2*OUTPUT_TERMINAL_DESC_SIZE),				      \
+	.bInCollection = ifaces,					      \
+	.baInterfaceNr = {0},						      \
+}
+
+#define INIT_IN_TERMINAL(dev, i, terminal_id, type)		\
+{								\
+	.bLength = INPUT_TERMINAL_DESC_SIZE,			\
+	.bDescriptorType = USB_CS_INTERFACE_DESC,		\
+	.bDescriptorSubtype = USB_AUDIO_INPUT_TERMINAL,		\
+	.bTerminalID = terminal_id,				\
+	.wTerminalType = sys_cpu_to_le16(type),			\
+	.bAssocTerminal = 0,					\
+	.bNrChannels = MAX(1, CH_CNT(dev, i)),			\
+	.wChannelConfig = sys_cpu_to_le16(CH_CFG(dev, i)),	\
+	.iChannelNames = 0,					\
+	.iTerminal = 0,						\
+}
+
+#define INIT_OUT_TERMINAL(terminal_id, source_id, type)		\
+{								\
+	.bLength = OUTPUT_TERMINAL_DESC_SIZE,			\
+	.bDescriptorType = USB_CS_INTERFACE_DESC,		\
+	.bDescriptorSubtype = USB_AUDIO_OUTPUT_TERMINAL,	\
+	.bTerminalID = terminal_id,				\
+	.wTerminalType = sys_cpu_to_le16(type),			\
+	.bAssocTerminal = 0,					\
+	.bSourceID = source_id,					\
+	.iTerminal = 0,						\
+}
+
+/** refer to Table 4-7 from audio10.pdf
+ */
+#define INIT_FEATURE_UNIT(dev, i, unit_id, source_id)			\
+{									\
+	.bLength = sizeof(struct dev##_feature_unit_descriptor_##i),	\
+	.bDescriptorType = USB_CS_INTERFACE_DESC,			\
+	.bDescriptorSubtype = USB_AUDIO_FEATURE_UNIT,			\
+	.bUnitID = unit_id,						\
+	.bSourceID = source_id,						\
+	.bControlSize = sizeof(u16_t),					\
+	.bmaControls = { FEATURES(dev, i) },				\
+	.iFeature = 0,							\
+}
+
+/* Class-Specific AS Interface Descriptor 4.5.2 audio10.pdf */
+#define INIT_AS_GENERAL(link)				\
+{							\
+	.bLength = USB_AC_CS_IF_DESC_SIZE,		\
+	.bDescriptorType = USB_CS_INTERFACE_DESC,	\
+	.bDescriptorSubtype = USB_AUDIO_AS_GENERAL,	\
+	.bTerminalLink = link,				\
+	.bDelay = 0,					\
+	.wFormatTag = sys_cpu_to_le16(0x0001),		\
+}
+
+/** Class-Specific AS Format Type Descriptor 4.5.3 audio10.pdf
+ * TODO: bFormatType
```

Comment:
I will remove this todo. I left this from the very beggining when I thaugh it would be easy to swap between formats (there are 3 in total). We are supporting format type I so it must be `0x01` and nothing more.